### DISCUSSION TITLE
Creating Roslyn code actions for formatting

### DISCUSSION THREAD

**Discussion by shocklateboy92** [dotnet/roslyn#54099] [Ideas]
### Proposal
I'm one of the maintainers of [csharpier](https://github.com/belav/csharpier), an complete code fromatter for C# inspired by JavaScript's [prettier](https://prettier.io/) and python's [black](https://github.com/psf/black).

While working out the best way to do IDE integration, we realized that Roslyn already contains
- Code formatting functionality, used by OmniSharp (and other plugins)
- A way for nuget packages to provide behavior at build/IDE time ([Roslyn analyzers / code fixes](https://docs.microsoft.com/en-us/visualstudio/code-quality/roslyn-analyzers-overview?view=vs-2019))

Putting the two together, I'd like to create a new type of code action / plugin. Roslyn will delegate formatting to any plugins used by a given project just for that project. 

### Motivation for approach
The biggest advantage of this approach (on top of all the [advantages an opinionated code formatter gives](https://prettier.io/docs/en/why-prettier.html)) is that the formatter and it's version will be tied to a project (via csproj/nuget reference). That way, IDEs will automatically switch between different styles for different projects, allowing the formatting to stay consistent.

### Next steps
I was told that all proposals for Rosyln/C# start as github discussions. I eagerly await feedback so I can flesh out the finer details.

**Comment by CyrusNajmabadi** [comment]
Tagging @sharwell :-)
